File: 02_algorithm/programmers/Level2/(2018-KAKAO-1차)뉴스클러스터링.py
def solution(str1, str2):
    if len(str1) == len(str2) == 0:
        return 65536
    elif not len(str1) or not len(str2):
        return 0
    else:
        # 1. 두 글자씩 글자쌍 나누기
        str_pair1, str_pair2 = [], []
        for idx_1 in range(len(str1) - 1):
            condition = [65 <= ord(str1[idx_1]) <= 90, 97 <= ord(str1[idx_1]) <= 122,
                         65 <= ord(str1[idx_1 + 1]) <= 90, 97 <= ord(str1[idx_1 + 1]) <= 122]
            if (condition[0] | condition[1]) & (condition[2] | condition[3]):
                str_pair1.append(str1[idx_1 : idx_1 + 2].lower())
        for idx_2 in range(len(str2) - 1):
            condition = [65 <= ord(str2[idx_2]) <= 90, 97 <= ord(str2[idx_2]) <= 122,
                         65 <= ord(str2[idx_2 + 1]) <= 90, 97 <= ord(str2[idx_2 + 1]) <= 122]
            if (condition[0] | condition[1]) & (condition[2] | condition[3]):
                str_pair2.append(str2[idx_2 : idx_2 + 2].lower())

        # 2. 두 리스트 합침 -> 집합으로 변환
        # -> for문으로 돌리면서 각 str_pair1, str_pair2에서 개수를 구한 후 최소값 = 교집합 개수에 더하고 최대값은 합집합 개수에 더함
        all_set = set(str_pair1 + str_pair2)
        intersection_cnt, union_cnt = 0, 0
        for element in all_set:
            min_value = min(str_pair1.count(element), str_pair2.count(element))
            max_value = max(str_pair1.count(element), str_pair2.count(element))
            intersection_cnt += min_value
            union_cnt += max_value
        if intersection_cnt == union_cnt == 0:
            return 65536
        return int((intersection_cnt / union_cnt) * 65536)

print(solution('FRANCE', 'french'))
print(solution('handshake', 'shake hands'))
print(solution('aa1+aa2', 'AAAA12'))
print(solution('E=M*C^2', 'e=m*c^2'))

File: 06_js/04_algorithm/programmers/Level2/(2018-KAKAO-1차)뉴스 클러스터링.js
function extractElement(originalString) {
  const reg = /[^a-zA-Z]/;
  let subSet = [];
  for (let i = 0; i < originalString.length - 1; i++) {
    if (!reg.test(originalString.slice(i, i + 2))) {
      subSet.push(originalString.slice(i, i + 2).toLowerCase());
    }
  }
  return subSet;
}

function solution(str1, str2) {
  // 교집합 원소 개수, 합집합 원소 개수
  let intersectionCount = 0;
  let unionCount = 0;
  // str1, str2 두 글자씩 묶어서 다중집합의 원소 추출
  const subSet1 = extractElement(str1);
  const subSet2 = extractElement(str2);
  // 두 다중집합 모두 원소가 없는 경우 65536을 return
  if (subSet1.length === 0 && subSet2.length === 0) {
    return 65536;
  }
  // 두 다중집합 중 하나의 집합 원소가 없는 경우 결과는 항상 0임
  if (subSet1.length === 0 || subSet2.length === 0) {
    return 0;
  }
  // 교집합, 합집합 계산
  let totalSubSet = new Set([...subSet1, ...subSet2]);
  for (let elem of totalSubSet) {
    const elementCount1 = subSet1.filter(e => e === elem).length;
    const elementCount2 = subSet2.filter(e => e === elem).length;
    intersectionCount += Math.min(elementCount1, elementCount2);
    unionCount += Math.max(elementCount1, elementCount2);
  }
  return parseInt((intersectionCount / unionCount) * 65536);
}

console.log(solution('FRANCE', 'french'));
console.log(solution('handshake', 'shake hands'));
console.log(solution('aa1+aa2', 'AAAA12'));
console.log(solution('E=M*C^2', 'e=m*c^2'));